### ISSUE TITLE
Point group of the layer group

### ISSUE THREAD

**Issue by site-g** [spglib/spglib#323]
The algorithm now requires the linear part **W** of the symmetry operation (**W**, **w**) to preserve the metric tensor **G**, i.e. **G = W<sup>T</sup>GW**. Should this be applied to the layer groups?

Should the non-lattice vector **c** affect the crystallographic point group? Or only the 2D lattice matters?

**Comment by lan496** [comment]
The condition $G = W^T G W$ is required for (W, w) to be isometry: that is, (W, w) does not change distances between two points. So, I think this condition still holds for layer groups.

**Comment by site-g** [comment]
I see. _**G = W**_<sup>T</sup>_**GW**_ should always hold.

I should modify my question to: If the input structure has layer group type _p_ 6 when the non-lattice vector **c** is perpendicular to the **ab** plane. Should it still be _p_ 6 when we distort **c** to arbitrary direction, but keep **a**, **b** and the cartesian coordinate of the atoms invariant?

**Comment by lan496** [comment]
I think it is also the case. The symmetry of the input structure itself does not depend on how to choose basis vectors $\mathbf{a}, \mathbf{b}, \mathbf{c}$. 
For your example, let $\mathbf{u}$ be a unit vector perpendicular to the $ab$ plane. A rotation part of the layer group operations transforms $\mathbf{u}$ to $\pm \mathbf{u}$. This property still holds for arbitrary scaled vector: $\lambda \mathbf{u}$ is also transformed to $\pm \lambda \mathbf{u}$.

**Comment by site-g** [comment]
Ok. I am just not sure if there are any crystallographic theory that requires **c** to affect the layer group.

Then the algorithm might need some modification. Now, only _**W**_ with elements 0, 1, -1 are searched. When the basis vector **c** is not perpendicular to the _ab_ plane, the operations with perpendicular axis along **u** cannot be represent by such matrices. Do you think we should add a step of projecting **c** to **u**?

**Comment by lan496** [comment]
Yes, the step to project $\mathbf{c}$ to the perpendicular axis will be clear in the algorithm perspective and make the implementation simpler.

**Comment by atztogo** [comment]
I don't see what your definition of projection is.

I think we can use the same routine as space group and then we can collect those  _**W**_ where **a** and **b** are kept in the same plane represented by original **a** and **b**. Following Table 1.2.1.1 of ITE, for  _p_ 6 case, if **c** + m**a** + n**b** (m and n are integers) can not make alpha=beta=90, we can see the symmetry of the user input is broken. What do you think?

**Comment by site-g** [comment]
@atztogo if my understanding is correct, your opinion is that the direction of **c** affects the found layer group. But **c** is only a basis vector of the coordinate system, it is not a lattice vector now. Sould it be able to break the symmetry now?

**Comment by atztogo** [comment]
I discuss about our definition. For example, in our definition, we can force users to choose a and b are lattice vectors. Under this setting, Table 1.2.1.1 of ITE mentions about alpha, beta, gamma, which means c can be not perpendicular to a-b plane. Then, we can define our rule so that the best new c is found by c + ma + nb. This is a propose.

**Comment by site-g** [comment]
I agree that **c** can be not perpendicular to _ab_ plane. Under this circustances, something like 6-fold rotation will be represented by a matrix with fractional numbers. What is the reason to drop them?

**Comment by atztogo** [comment]
W's are the candidates of symmetry operations. Symmetry operations are defined with respect to a, b, c for layer group like space group although c is the non lattice vector according to ITE. Therefore fractional numbers should be rejected.

**Comment by site-g** [comment]
This is the point that concerns me. I cannot find relative restrictions in ITE. What I found is that the conventional cell of layer group is 2D. So, I think the first two rows in **W** is guaranteed to be integer under the conventional coordinate system, but the number type of the third row is not defined?

**Comment by atztogo** [comment]
Please do not copy and paste from books. We have to respect copyright.
Rotation part of symmetry operation should be represented by an integer matrix.

**Comment by site-g** [comment]
I removed it.

**Comment by site-g** [comment]
Rotation part of symmetry operation should be represented by an integer matrix when the basis vector of this matrix are primitive or conventional cell basis vectors. However for layer group, ITE said the cell should be 2D.

**Comment by atztogo** [comment]
Lattice should be 2D, but a, b, c are defined if I understand correctly.

**Comment by site-g** [comment]
I think c is not defined, as you can see in the seventh column of Table 1.2.1.1 in ITE.

**Comment by atztogo** [comment]
I mean section 1.2.1.2.1 of ITE.

**Comment by site-g** [comment]
Section 1.2.1.2.1 defines **a**, **b** and **c** are the crystallographic basis of the crystallographic coordinate system. What I do not know is why would a coordinate system add restrictions on the symmetry group type of the structure? I think only the lattice or the cell should do this, because they affect the translation subgroup of the layer group.

**Comment by atztogo** [comment]
It is not a question to me, but I think it is what we define for our convenience.

**Comment by site-g** [comment]
I see. So your opinion is that we should preserve the current algorithm, i.e. drop the rotation if it cannot be represented by an integer matrix under the basis vector **a**, **b** and **c**. If the users want higher symmetry operations, they should make **c** vertical to the _ab_ plane themselves. Is my understanding correct?

**Comment by lan496** [comment]
> It is not a question to me, but I think it is what we define for our convenience.

I agree as well. Table 1.2.1.1 specifies a conventional choice of basis vectors.

In my understanding, the current implementation assumes that a user's input basis vectors $(a', b', c')$ satisfies
(1) $a'$ and $b'$ generate a subgroup of a translation subgroup of the layer group.
And, we do not assume
(2) $c'$ is perpendicular to $a'b'$ plane.

If both of the following two conditions hold, symmetry operations of the layer group should be represented with an integer matrix in $(a', b', c')$ (I explain next): 
(i) $a'$ and $b'$ are primitive basis vectors for the translation subgroup of the layer group
(ii) let $c$ be a perpendicular vector to the ab plane. $c'$ can be written as $c' = c + ma + nb$ ($m$, $n$ are integers).

Under (i), $(a', b', c)$ provides an integer matrix $W$ for each symmetry operation.
Under (ii), the user's basis vectors $(a', b', c')$ and $(a', b', c)$ is transformed to the other by an unimodular matrix,

$$
P = \begin{pmatrix}
1 & 0 & m \\
0 & 1 & n \\
0 & 0 & 1 \\
\end{pmatrix}.
$$

Thus, $P^{-1} W P$ is also an integer matrix (because $P^{-1}$ is an integer matrix).

Indeed, if the condition (i) or (ii) does not hold, the matrix may have fractional values.
In that case, I think we should adopt the convention to drop it because similar behavior is already used in spglib's space group search described in the last paragraph of Appendix C in [spglib's arxiv paper](https://arxiv.org/pdf/1808.01590.pdf)

**Comment by site-g** [comment]
So, the most important is condition (ii), which is a Spglib convention.


> Appendix C in [spglib's arxiv paper](https://arxiv.org/pdf/1808.01590.pdf)

I think this is done after the space group type has been found. For example, _P_ 6 has been found, but the input cell is a 2x1x1 supercell of the conventional cell that breaks the symmetry. Then the 6-fold rotation will be dropped, but Spglib still output the space group number as 168, as documented in [spg_get_symmetry](https://spglib.readthedocs.io/en/latest/api.html#spg-get-symmetry).

**Comment by atztogo** [comment]
@lan496's condition (1) is the rule that we ask users, but the condition (2) is not necessarily hold. Before the point group search against a,b,c, we have the reduced cell routine, where we may use 2D Delaunay reduction for a and b combination (-> a', b'). For c, we probably look for a shortest by c'=c+ma'+nb'. For P6 case, c' finally should be perpendicular to a-b plane, otherwise user must accept P6 symmetry is broken.

@site-g's last comment is correction. But I guess @lan496 means just that we can use the same login to reject operations. For space group case, this comes to last step, because spglib algorithm first determines a primitive cell and the space group is searched for the primitive cell. For supercell, the symmetry operations of primitive cell are expanded by the translation of lattice points in the supercell. Then the check explained by @lan496 is used to reject some symmetry operations for the supercell. This is a tricky part because we implicitly define supercell symmetry that is incompatible with the coset decomposition of the space group but for the practical purpose (e.g. for using the symmetry operations for electronic structure calculation) it is necessary.

**Comment by site-g** [comment]
> This is a tricky part because we implicitly define supercell symmetry that is incompatible with the coset decomposition of the space group but for the practical purpose (e.g. for using the symmetry operations for electronic structure calculation) it is necessary.

I had some difficulty in understanding this sentence. Would you mind give a more concrete example?

**Comment by atztogo** [comment]
This is not the argument of the crystallography, a kind of convention of the electronic structure calculation community. I don't like to discuss about this, because this is just so defined in the many codes. So it is not very much logical. As a convention, if the supercell shape breaks the point group of the primitive cell, the broken symmetry of the sueprcell lattice is used as the initial set of the rotations for searching space group operations. This results in, if your input crystal structure is a supercell but the lattice formed by the supercell basis vectors is different from the point group of the space group type, the symmetry operations found for the supercell will be different from the symmetry operations of the primitive cell times the lattice translations in the supercell with respect to the primitive cell.

**Comment by site-g** [comment]
I see. So, the convention is following the convention of many other codes. While the difference is that, they usually drop the operations at the start stage, Spglib drops the operations at the end stage to mimic the behavior of the other codes.
